internal_task_handlers.go

```diff
@@ -308,83 +324,242 @@ func newWorkflowTaskHandler(
+	return w.eventHandler == nil
+}
+
+func (w *workflowExecutionContext) destroyCachedState() {
+	w.isWorkflowCompleted = false
+	w.result = nil
+	w.err = nil
+	w.previousStartedEventID = 0
+	if w.eventHandler != nil {
+		w.eventHandler.Close()
+		w.eventHandler = nil
+	}
+}
+
+func (w *workflowExecutionContext) resetWorkflowState() {
+	w.destroyCachedState()
+	w.eventHandler = newWorkflowExecutionEventHandler(
+		w.workflowInfo,
+		w.completeWorkflow,
+		w.wth.logger,
+		w.wth.enableLoggingInReplay,
+		w.wth.metricsScope,
+		w.wth.hostEnv)
+}
+
+func resetHistory(task *s.PollForDecisionTaskResponse, historyIterator HistoryIterator) (*s.History, error) {
+	historyIterator.Reset()
+	firstPageHistory, err := historyIterator.GetNextPage()
+	if err != nil {
+		return nil, err
 	}
-	attributes := event.WorkflowExecutionStartedEventAttributes
+	task.History = firstPageHistory
+	return firstPageHistory, nil
+}
+
+func (wth *workflowTaskHandlerImpl) createWorkflowContext(task *s.PollForDecisionTaskResponse) (*workflowExecutionContext, error) {
+	h := task.History
+	attributes := h.Events[0].WorkflowExecutionStartedEventAttributes
 	if attributes == nil {
-		return nil, "", errors.New("first history event is not WorkflowExecutionStarted")
+		return nil, errors.New("first history event is not WorkflowExecutionStarted")
 	}
 	taskList := attributes.TaskList
 	if taskList == nil {
-		return nil, "", errors.New("nil TaskList in WorkflowExecutionStarted event")
+		return nil, errors.New("nil TaskList in WorkflowExecutionStarted event")
 	}
 
-	traceLog(func() {
-		wth.logger.Debug("Processing new workflow task.",
-			zap.String(tagWorkflowType, task.WorkflowType.GetName()),
-			zap.String(tagWorkflowID, task.WorkflowExecution.GetWorkflowId()),
-			zap.String(tagRunID, task.WorkflowExecution.GetRunId()),
-			zap.Int64("PreviousStartedEventId", task.GetPreviousStartedEventId()))
-	})
+	runID := task.WorkflowExecution.GetRunId()
+	workflowID := task.WorkflowExecution.GetWorkflowId()
 
 	// Setup workflow Info
 	workflowInfo := &WorkflowInfo{
 		WorkflowType: flowWorkflowTypeFrom(*task.WorkflowType),
 		TaskListName: taskList.GetName(),
 		WorkflowExecution: WorkflowExecution{
-			ID:    *task.WorkflowExecution.WorkflowId,
-			RunID: *task.WorkflowExecution.RunId,
+			ID:    workflowID,
+			RunID: runID,
 		},
 		ExecutionStartToCloseTimeoutSeconds: attributes.GetExecutionStartToCloseTimeoutSeconds(),
 		TaskStartToCloseTimeoutSeconds:      attributes.GetTaskStartToCloseTimeoutSeconds(),
 		Domain: wth.domain,
 	}
+	wfStartTime := time.Unix(0, h.Events[0].GetTimestamp())
+	workflowContext := &workflowExecutionContext{workflowStartTime: wfStartTime, workflowInfo: workflowInfo, wth: wth}
+	workflowContext.resetWorkflowState()
 
-	isWorkflowCompleted := false
-	var completionResult []byte
-	var failure error
+	return workflowContext, nil
+}
+
+func (wth *workflowTaskHandlerImpl) getOrCreateWorkflowContext(task *s.PollForDecisionTaskResponse,
```

Comment:
While I couldn't find any obvious problems with this method I believe it could be refactored to improve it readability.

Reply:
the problem is the query is not sticky for now which make it more complex than it needs to be.